src/common/Analytics/AnalyticsEvents.ts

```diff
@@ -0,0 +1,70 @@
+import { EventEmitter } from 'events';
+
+export type AnalyticsEventTypeHandlers = {
+  track: (eventName: string, event: unknown) => void;
+  userPropertyUpdate: (
+    key: string,
+    value: string | number | boolean | undefined,
+  ) => void;
+  reset: () => void;
+};
+
+export type AnalyticsEventTypes = keyof AnalyticsEventTypeHandlers;
+export type AnalyticsEventTypeHandler<T extends AnalyticsEventTypes> =
+  AnalyticsEventTypeHandlers[T];
+
+export class AnalyticsEvents {
+  private emitter = new EventEmitter();
+
+  public addListener<EventType extends AnalyticsEventTypes>(
+    eventType: EventType,
+    listener: AnalyticsEventTypeHandler<EventType>,
+  ) {
+    return this.emitter.addListener(eventType, listener);
+  }
+
+  public removeListener<EventType extends AnalyticsEventTypes>(
+    eventType: EventType,
+    listener: AnalyticsEventTypeHandler<EventType>,
+  ) {
+    return this.emitter.removeListener(eventType, listener);
+  }
+
+  public emit<EventType extends AnalyticsEventTypes>(
+    eventType: EventType,
+    ...params: Parameters<AnalyticsEventTypeHandler<EventType>>
+  ) {
+    return this.emitter.emit(eventType, ...params);
+  }
+}
+
+export const analyticsEvents = new AnalyticsEvents();
+
+export type Tracker<CustomEventMap extends Record<string, unknown>> = {
+  track<Key extends keyof CustomEventMap>(
+    name: Key,
+    event: CustomEventMap[Key],
+  ): void;
+  userPropertyUpdate: AnalyticsEventTypeHandlers['userPropertyUpdate'];
+  reset: AnalyticsEventTypeHandlers['reset'];
```

Comment:
instead of explicitly adding these here, could we just do `& AnalyticsEventTypeHandlers`? We could then just override `track`.

Reply:
Also, looking at this more closely, I think we could do with a single generic type/interface, right? Instead of having both `AnalyticsEventTypeHandlers` and `Tracker`?

Reply:
I was not able to get the types to play nicely with a single handler/tracker object.